src/java/org/apache/cassandra/io/tries/IncrementalDeepTrieWriterPageAware.java

```diff
@@ -0,0 +1,405 @@
+        {
+            if (child.filePos == -1)
+            {
+                childrenToClear.add(child);
+                if (depth < maxRecursionDepth)
+                    child.filePos = writePartialRecursiveOnStack(child, dest, baseOffset, depth + 1);
+                else
+                    child.filePos = writePartialRecursiveOnHeap(child, dest, baseOffset);
+            }
+        }
+
+        long nodePosition = dest.position() + baseOffset;
+
+        if (node.hasOutOfPageInBranch)
+        {
+            // Update the branch size with the size of what we have just written. This may be used by the node's
+            // maxPositionDelta, and it's a better approximation for later fitting calculations.
+            node.branchSize = (int) (nodePosition - startPosition);
+        }
+
+        serializer.write(dest, node, nodePosition);
+
+        if (node.hasOutOfPageChildren || node.hasOutOfPageInBranch)
+        {
+            // Update the node size with what we have just seen. It's a better approximation for later fitting
+            // calculations.
+            long endPosition = dest.position() + baseOffset;
+            node.nodeSize = (int) (endPosition - nodePosition);
+        }
+
+        for (Node<VALUE> child : childrenToClear)
+            child.filePos = -1;
+        return nodePosition;
+    }
+
+    private long writePartialRecursiveOnHeap(Node<VALUE> node, DataOutputPlus dest, long baseOffset) throws IOException
+    {
+        new WritePartialRecursion(node, dest, baseOffset).process();
+        long pos = node.filePos;
+        node.filePos = -1;
+        return pos;
+    }
+
+    class WritePartialRecursion extends Recursion<Node<VALUE>>
+    {
+        final DataOutputPlus dest;
+        final long baseOffset;
+        final long startPosition;
+        final List<Node<VALUE>> childrenToClear;
+
+        WritePartialRecursion(Node<VALUE> node, WritePartialRecursion parent)
+        {
+            super(node, node.children.iterator(), parent);
+            this.dest = parent.dest;
+            this.baseOffset = parent.baseOffset;
+            this.startPosition = dest.position() + baseOffset;
+            childrenToClear = new ArrayList<>();
+        }
+
+        WritePartialRecursion(Node<VALUE> node, DataOutputPlus dest, long baseOffset)
+        {
+            super(node, node.children.iterator(), null);
+            this.dest = dest;
+            this.baseOffset = baseOffset;
+            this.startPosition = dest.position() + baseOffset;
+            childrenToClear = new ArrayList<>();
+        }
+
+        @Override
+        Recursion<Node<VALUE>> makeChild(Node<VALUE> child)
+        {
+            if (child.filePos == -1)
+            {
+                childrenToClear.add(child);
+                return new WritePartialRecursion(child, this);
+            }
+            else
+                return null;
+        }
+
+        @Override
+        void complete() throws IOException
```

Comment:
nit: Seems like the duplication is intended, so maybe throw a `@SuppressWarnings("DuplicatedCode")` on there.